### ISSUE TITLE
NullPointerException cleaning tests

### ISSUE THREAD

**Issue by lantoli** [concordion/concordion-logback-extension#5]
Hi,
when I execute a test with run commands with Eclipse I get an exception when the test is finishing. However I don't get the exception when executing with Maven.
Don't know why it's different if executed in Eclipse or Maven.
I'm using Concordion 1.5.0-snapshot.

The error is because name is null in:

public static String getTestClassName() {
                               String name = MDC.get(TEST_NAME);
 
                               return name.substring(name.lastIndexOf("/") + 1);
                }
 

Trace:

ene 09, 2015 12:59:52 PM org.concordion.internal.runner.DefaultConcordionRunner logExceptionIfNotAssertionError
WARNING:
java.lang.NullPointerException
                at org.concordion.ext.loggingFormatter.LogbackHelper.getTestClassName(LogbackHelper.java:112)
                at org.concordion.ext.loggingFormatter.LoggingFormatterSpecificationListener.createViewer(LoggingFormatterSpecificationListener.java:71)
                at org.concordion.ext.loggingFormatter.LoggingFormatterSpecificationListener.afterProcessingSpecification(LoggingFormatterSpecificationListener.java:40)
                at sun.reflect.GeneratedMethodAccessor8.invoke(Unknown Source)
                at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                at java.lang.reflect.Method.invoke(Method.java:483)
                at org.concordion.internal.util.Announcer.announce(Announcer.java:47)
                at org.concordion.internal.util.Announcer.access$000(Announcer.java:16)
                at org.concordion.internal.util.Announcer$1.invoke(Announcer.java:26)
                at com.sun.proxy.$Proxy10.afterProcessingSpecification(Unknown Source)
                at org.concordion.internal.command.SpecificationCommand.announceAfterProcessingEvent(SpecificationCommand.java:59)
                at org.concordion.internal.command.SpecificationCommand.execute(SpecificationCommand.java:35)
                at org.concordion.api.CommandCall.execute(CommandCall.java:29)
                at org.concordion.internal.XMLSpecification.process(XMLSpecification.java:17)
                at org.concordion.Concordion.process(Concordion.java:32)
                at org.concordion.Concordion.process(Concordion.java:26)
                at org.concordion.internal.FixtureRunner.run(FixtureRunner.java:13)
                at org.concordion.integration.junit4.ConcordionRunner$1.evaluate(ConcordionRunner.java:113)
                at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
                at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
                at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
                at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
                at org.concordion.integration.junit4.ConcordionRunner.runChild(ConcordionRunner.java:104)
                at org.concordion.integration.junit4.ConcordionRunner.runChild(ConcordionRunner.java:18)
                at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
                at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
                at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
                at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
                at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
                at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
                at org.junit.runners.Suite.runChild(Suite.java:127)
                at org.junit.runners.Suite.runChild(Suite.java:26)
                at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
                at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
                at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
                at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
                at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
                at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
                at org.junit.runner.JUnitCore.run(JUnitCore.java:160)
                at org.junit.runner.JUnitCore.run(JUnitCore.java:138)
                at org.junit.runner.JUnitCore.run(JUnitCore.java:128)
                at org.junit.runner.JUnitCore.runClasses(JUnitCore.java:73)
                at org.concordion.internal.runner.DefaultConcordionRunner.runJUnitClass(DefaultConcordionRunner.java:57)
                at org.concordion.internal.runner.DefaultConcordionRunner.runTestClass(DefaultConcordionRunner.java:52)
                at org.concordion.internal.runner.DefaultConcordionRunner.execute(DefaultConcordionRunner.java:22)
                at org.concordion.internal.command.SequentialRunStrategy.call(SequentialRunStrategy.java:13)
                at org.concordion.internal.command.RunCommand.execute(RunCommand.java:101)
                at org.concordion.api.AbstractCommandDecorator$2.run(AbstractCommandDecorator.java:23)
                at org.concordion.internal.command.LocalTextDecorator.process(LocalTextDecorator.java:26)
                at org.concordion.api.AbstractCommandDecorator.execute(AbstractCommandDecorator.java:21)
                at org.concordion.api.AbstractCommandDecorator$2.run(AbstractCommandDecorator.java:23)
                at org.concordion.internal.command.ThrowableCatchingDecorator.process(ThrowableCatchingDecorator.java:42)
                at org.concordion.api.AbstractCommandDecorator.execute(AbstractCommandDecorator.java:21)
                at org.concordion.api.CommandCall.execute(CommandCall.java:29)
                at org.concordion.api.CommandCallList.processSequentially(CommandCallList.java:30)
                at org.concordion.internal.command.SpecificationCommand.execute(SpecificationCommand.java:31)
                at org.concordion.api.CommandCall.execute(CommandCall.java:29)
                at org.concordion.internal.XMLSpecification.process(XMLSpecification.java:17)
                at org.concordion.Concordion.process(Concordion.java:32)
                at org.concordion.Concordion.process(Concordion.java:26)
                at org.concordion.internal.FixtureRunner.run(FixtureRunner.java:13)
                at org.concordion.integration.junit4.ConcordionRunner$1.evaluate(ConcordionRunner.java:113)
                at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
                at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
                at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
                at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
                at org.concordion.integration.junit4.ConcordionRunner.runChild(ConcordionRunner.java:104)
                at org.concordion.integration.junit4.ConcordionRunner.runChild(ConcordionRunner.java:18)
                at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
                at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
                at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
                at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
                at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
                at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
                at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:50)
                at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
                at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
                at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:675)
                at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
                at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)

**Comment by nigelcharman** [comment]
With the parallel runner, I wonder if this code could be executing on a different thread. Would you try reducing the thread count to 1 and see if it still occurs? Maybe you have different thread counts in Maven and Eclipse? 



On 9 January 2015 1:53:03 PM GMT+00:00, Leo Antoli <notifications@github.com> wrote:
>Hi,
>when I execute a test with run commands with Eclipse I get an exception
>when the test is finishing. However I don't get the exception when
>executing with Maven.
>Don't know why it's different if executed in Eclipse or Maven.
>I'm using Concordion 1.5.0-snapshot.
>
>The error is because name is null in:
>
>public static String getTestClassName() {
>                               String name = MDC.get(TEST_NAME);
> 
>                      return name.substring(name.lastIndexOf("/") + 1);
>                }
> 
>
>Trace:
>
>ene 09, 2015 12:59:52 PM
>org.concordion.internal.runner.DefaultConcordionRunner
>logExceptionIfNotAssertionError
>WARNING:
>java.lang.NullPointerException
>at
>org.concordion.ext.loggingFormatter.LogbackHelper.getTestClassName(LogbackHelper.java:112)
>at
>org.concordion.ext.loggingFormatter.LoggingFormatterSpecificationListener.createViewer(LoggingFormatterSpecificationListener.java:71)
>at
>org.concordion.ext.loggingFormatter.LoggingFormatterSpecificationListener.afterProcessingSpecification(LoggingFormatterSpecificationListener.java:40)
>         at sun.reflect.GeneratedMethodAccessor8.invoke(Unknown Source)
>at
>sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
>                at java.lang.reflect.Method.invoke(Method.java:483)
>  at org.concordion.internal.util.Announcer.announce(Announcer.java:47)
>at org.concordion.internal.util.Announcer.access$000(Announcer.java:16)
>  at org.concordion.internal.util.Announcer$1.invoke(Announcer.java:26)
> at com.sun.proxy.$Proxy10.afterProcessingSpecification(Unknown Source)
>at
>org.concordion.internal.command.SpecificationCommand.announceAfterProcessingEvent(SpecificationCommand.java:59)
>at
>org.concordion.internal.command.SpecificationCommand.execute(SpecificationCommand.java:35)
>         at org.concordion.api.CommandCall.execute(CommandCall.java:29)
>at
>org.concordion.internal.XMLSpecification.process(XMLSpecification.java:17)
>               at org.concordion.Concordion.process(Concordion.java:32)
>               at org.concordion.Concordion.process(Concordion.java:26)
>    at org.concordion.internal.FixtureRunner.run(FixtureRunner.java:13)
>at
>org.concordion.integration.junit4.ConcordionRunner$1.evaluate(ConcordionRunner.java:113)
>at
>org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
>at
>org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
>       at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
>at
>org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
>at
>org.concordion.integration.junit4.ConcordionRunner.runChild(ConcordionRunner.java:104)
>at
>org.concordion.integration.junit4.ConcordionRunner.runChild(ConcordionRunner.java:18)
>         at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
>     at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
>   at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
>     at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
>    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
>           at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
>                at org.junit.runners.Suite.runChild(Suite.java:127)
>                at org.junit.runners.Suite.runChild(Suite.java:26)
>         at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
>     at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
>   at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
>     at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
>    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
>           at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
>                at org.junit.runner.JUnitCore.run(JUnitCore.java:160)
>                at org.junit.runner.JUnitCore.run(JUnitCore.java:138)
>                at org.junit.runner.JUnitCore.run(JUnitCore.java:128)
>            at org.junit.runner.JUnitCore.runClasses(JUnitCore.java:73)
>at
>org.concordion.internal.runner.DefaultConcordionRunner.runJUnitClass(DefaultConcordionRunner.java:57)
>at
>org.concordion.internal.runner.DefaultConcordionRunner.runTestClass(DefaultConcordionRunner.java:52)
>at
>org.concordion.internal.runner.DefaultConcordionRunner.execute(DefaultConcordionRunner.java:22)
>at
>org.concordion.internal.command.SequentialRunStrategy.call(SequentialRunStrategy.java:13)
>at
>org.concordion.internal.command.RunCommand.execute(RunCommand.java:101)
>at
>org.concordion.api.AbstractCommandDecorator$2.run(AbstractCommandDecorator.java:23)
>at
>org.concordion.internal.command.LocalTextDecorator.process(LocalTextDecorator.java:26)
>at
>org.concordion.api.AbstractCommandDecorator.execute(AbstractCommandDecorator.java:21)
>at
>org.concordion.api.AbstractCommandDecorator$2.run(AbstractCommandDecorator.java:23)
>at
>org.concordion.internal.command.ThrowableCatchingDecorator.process(ThrowableCatchingDecorator.java:42)
>at
>org.concordion.api.AbstractCommandDecorator.execute(AbstractCommandDecorator.java:21)
>         at org.concordion.api.CommandCall.execute(CommandCall.java:29)
>at
>org.concordion.api.CommandCallList.processSequentially(CommandCallList.java:30)
>at
>org.concordion.internal.command.SpecificationCommand.execute(SpecificationCommand.java:31)
>         at org.concordion.api.CommandCall.execute(CommandCall.java:29)
>at
>org.concordion.internal.XMLSpecification.process(XMLSpecification.java:17)
>               at org.concordion.Concordion.process(Concordion.java:32)
>               at org.concordion.Concordion.process(Concordion.java:26)
>    at org.concordion.internal.FixtureRunner.run(FixtureRunner.java:13)
>at
>org.concordion.integration.junit4.ConcordionRunner$1.evaluate(ConcordionRunner.java:113)
>at
>org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
>at
>org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
>       at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
>at
>org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
>at
>org.concordion.integration.junit4.ConcordionRunner.runChild(ConcordionRunner.java:104)
>at
>org.concordion.integration.junit4.ConcordionRunner.runChild(ConcordionRunner.java:18)
>         at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
>     at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
>   at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
>     at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
>    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
>           at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
>at
>org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:50)
>at
>org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
>at
>org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
>at
>org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:675)
>at
>org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
>at
>org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)
> 
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/concordion/concordion-logback-extension/issues/5

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by lantoli** [comment]
I think I've found when NullPointerException happens, it is when concordion.run.threadCount is not defined.   
if it is defined, it works fine, be 1, 5, or whatever value.   But I always get NullPointerException if the variable is not defined.
So apparently there is a difference between not defining the variable and defining it with value 1. 
As a matter of fact in the latter case I get the log message: "Running concordion:run commands in parallel with 1 threads".

This exception only happens in specs with run commands.

I hope it helps. Thanks.

**Comment by nigelcharman** [comment]
Hi, I've been unable to reproduce the issue, other than by not calling `LogbackHelper.startTestLogging()`, or moving `stopTestLogging()` to occur immediately after `startTestLogging()`. Do you have `stopTestLogging()` in your `@After` block? Maybe you could try removing the calls to `stopTestLogging()` to see what effect it has?

Not defining the `concordion.run.threadCount` variable causes the code to use the existing serial runner, rather than the parallel runner. I can't currently see why that would be making a difference.

Can you provide a minimal test case please?

@andrew-sumner I suggest you add defensive code to `getTestClassName()` in case the key is not found. However, I don't think it will resolve this problem (other than changing the NPE into an empty log file).

**Comment by andrew-sumner** [comment]
I've managed to duplicate the issue accidentally at work today.  I've added an example to the logback-extension in src/test/java/temp/LogbackLoggingIndex.java.

The issue is that I am using MDC to split the tests - but that only works if tests are running on different threads.  So the minute you have a test calling another test with run command things start to get a little screwy.

I've never encountered this before because usually I'm running multi-threaded and/or the only time I use the run command is in an index and I don't have the logback extension switched on for indexes.

I'm open to ideas on a more robust way to split logs into a separate log per test using logback that doesn't use MDC...
